gpui: Dynamic element arena

MatinAniss opened the pull request:
Implements a chunking strategy for the element arena that allows it to grow dynamically based on allocations, it is initialised with a single chunk of a total size of 1 mebibyte. On allocation of data with a size greater than the remaining space of the current chunk a new chunk is created.

This reduces the memory allocation from the static 32 mebibytes, this especially helps GPUI applications that don't need such a large element arena and even Zed in most cases. This also prevents the panic when allocations ever exceed the element arena.

Release Notes:

- N/A

MatinAniss committed:
gpui: Dynamic element arena

MatinAniss committed:
Add extra size check

filipwiech commented on crates/gpui/src/arena.rs:
```diff
@@ -79,30 +121,55 @@ impl Arena {
+            let mut current_chunk = self.chunks.get_mut(self.current_chunk).unwrap();
+            let ptr = if let Some(ptr) = current_chunk.allocate(layout) {
+                ptr
+            } else if self.current_chunk + 1 < self.chunks.len() {
```
Thanks to your explanation in https://github.com/zed-industries/zed/pull/32079#discussion_r2126896616 now I see that the chunk can be reused. :+1:

filipwiech commented on crates/gpui/src/arena.rs:
```diff
@@ -79,30 +121,53 @@ impl Arena {
+            let mut current_chunk = self.chunks.get_mut(self.current_chunk).unwrap();
+            let ptr = if let Some(ptr) = current_chunk.allocate(layout) {
+                ptr
+            } else if self.current_chunk + 1 < self.chunks.len()
```
2 points, but please ignore if they seem like a nitpick, I'm really not sure how important they are, if at all:
* Would be nice to add a test case for this scenario.
* In some situations we may end up with an unexpected (?) behaviour if we need to skip a chunk. Example:
  Arena after growing and getting cleared: [1 MB] -> [1 MB] -> [1 MB] -> [1MB]
  Now we need to allocate 2MB: [1 MB] -> [1 MB] -> [1 MB] -> [1MB] -> [2 MB]
  All the 1MB chunks are left unused.

mgsloan committed:
Merge branch 'main' into dynamic-arena

mgsloan committed:
Panic when attempting to allocate an element bigger than chunk_size

The previous code was buggy in this case. It would allocate a large element into the last chunk, but only increment `current_chunk`. This means that the large element's memory could be overwritten by later allocations.

mgsloan committed:
Remove per-chunk size tracking + Drop for Chunk

mgsloan committed:
Only reset chunks that need reset

mgsloan committed:
Check `alloc` for null result

mgsloan committed:
Assert nonzero chunk_size

mgsloan committed:
Bring back info log about arena capacity increases
